### PR TITLE
Add decorator to remove namespace from ClusterRole and ClusterRoleBin…

### LINKED ISSUE #43464: `quarkus.kubernetes.namespace` should not be applied to clustered resources
### Describe the bug

If you use the `quarkus.kubernetes.namespace` property, it will get applied indiscriminately to all generated Kubernetes resources regardless of whether they actually can accept a namespace or not. While this usually doesn't cause issues with `kubectl apply`, other processes might be more regarding with respect to structure validation. 

### Expected behavior

Clustered resources should not have a namespace.

### Actual behavior

All generated resources, including clustered ones, have their namespace field set.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mcruzdev**
- Fixes: #43464

### CODE REVIEW

**Comment by metacosm** [comment]
While this solution probably works, I would rather we don't add the namespace to clustered resources in the first place instead of adding it and then removing it.

**Comment by metacosm** [comment]
Sorry to be a pain but I think the `AddNamespaceDecorator` should actually be rewritten completely to deal with this instead of trying to order things in such a way that things work. Maybe the decorator should only target resources that implement `Namespaced` so that we know for sure that processed resources need a namespace?

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java [hunk 1]
```diff
@@ -43,4 +39,24 @@ public boolean equals(Object o) {
...
+        List<HasMetadata> collect = list.buildItems()
+                .stream()
+                .filter(o -> o instanceof HasMetadata)
+                .peek(o -> {
+                    if (eligibleForChangingNamespace(o)) {
+                        o.setMetadata(o.getMetadata().edit().withNamespace(namespace).build());
+                    }
+                }).collect(Collectors.toList());
+
+        list.removeAllFromItems(copy);
```
[the remark this answers is not in the document]
Comment (metacosm) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java:55]: I would do something like:
```java
final var updated = list.buildItems().stream()
                .filter(Namespaced.class::isInstance)
                .peek(o -> o.getMetadata().edit().withNamespace(namespace).build())
                .toList();
list.withItems(updated);
```
but I haven't checked if this actually works. This has the advantage to not require maintaining a list of clustered resources. Ideally, we wouldn't need to build the items in the first place but I'm not sure this is avoidable. `withItems` resets the list completely so it should be faster than first removing then adding the items again. Maybe @iocanel has a better solution?

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java [hunk 2]
```diff
@@ -43,4 +38,22 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(namespace);
     }
+
+    @Override
+    public void visit(KubernetesListBuilder list) {
+        List<HasMetadata> buildItems = list.buildItems()
+                .stream()
+                .filter(o -> o instanceof HasMetadata)
```
Comment (metacosm) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java:46]: Look at the code I sent previously, you should be able to simplify this even more and remove the `clusteredResources` set completely, if I'm not mistaken.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java]: I'm surprised that it doesn't work. I'll check tomorrow.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java [hunk 3]
```diff
@@ -1,27 +1,39 @@
...
 
     @Override
-    public void andThenVisit(ObjectMetaBuilder builder, ObjectMeta resourceMeta) {
-        if (!builder.hasNamespace()) {
-            builder.withNamespace(namespace);
-        }
+    public void visit(KubernetesListBuilder list) {
+        List<HasMetadata> buildItems = list.buildItems()
+                .stream()
+                .filter(obj -> obj instanceof HasMetadata)
```
Comment (metacosm) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java:26]: At the very least, you don't need this: all the items in the list should be `HasMetadata` instances.

File: extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java [hunk 4]
```diff
@@ -18,10 +20,19 @@ public AddNamespaceDecorator(String namespace) {
...
-        if (!builder.hasNamespace()) {
-            builder.withNamespace(namespace);
-        }
+    public void visit(KubernetesListBuilder list) {
+        List<HasMetadata> buildItems = list.buildItems()
+                .stream()
+                .peek(obj -> {
+                    if (obj instanceof Namespaced) {
+                        final ObjectMeta metadata = obj.getMetadata();
+                        if (metadata.getNamespace() == null) {
```
Comment (metacosm) [inline extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/AddNamespaceDecorator.java:29]: I forgot to add a comment here: this is meant to keep the previous functionality of not overriding a manually set namespace, i.e. the namespace provided via `quarkus.kubernetes.namespace` should only be applied to resources that don't already specify one.

**Review by metacosm — APPROVED** [review]

**Comment by metacosm** [comment]
The PR should be renamed more appropriately as well, since the PR is not adding a new decorator, simply fixing the existing one.

**Review by iocanel — COMMENTED** [review]
I really don't like it when a visitor operates on levels lower than its target, as it makes it really hard to order things.

I am wondering if it would be possible to crete a new kind of decroator based on sundrio's `PathAwareTypedVisitor` that allow devs to look up the parent node and / or the parent type. 

If this is not possible, then we shall proceed with this approach.

**Comment by metacosm** [comment]
> I really don't like it when a visitor operates on levels lower than its target, as it makes it really hard to order things.

I'm not sure I understand what you mean here.

> 
> I am wondering if it would be possible to crete a new kind of decroator based on sundrio's `PathAwareTypedVisitor` that allow devs to look up the parent node and / or the parent type.

That would require a new Dekorate release, though. And I'm not sure what benefits that would bring. How looking up the parent would help in this case? Do you mean to look up the resource object from its `ObjectMeta`?
The current approach seems simpler with the downside of requiring building the objects instead of working directly on builders.

**Review by iocanel — APPROVED** [review]
The suggested change not only requires a decorate release, but it requires a major release that is also not trivial.

So, given that it will require time, I am going to accept this.

**Comment by metacosm** [comment]
@mcruzdev I agree but this isn't my call to make 🙍🏼

**Review by cescoffier — APPROVED** [review]

**Pull request merged**